### PR TITLE
Online Diagnostics Module

### PR DESCRIPTION
**Pull request by LenkaNovak**
## Purpose 

Modularize the diagnostics utility. 

## To-do

- Tasks:
  - [x] move functions to src
  - [x] allow instantaneous save
  - [x] add tests
  - [x] add docs
  - [x] update Utlilities
  - [x] wrap into module after testing
  - [x] update TimeManager with with `trigger_callback` (for better tracking of callback types) and `Monthly()`

## Note
- here we still use ClimaComms 0.3.2, to enable other PRs to go through and avoid dependency problems with ClimaCore (https://buildkite.com/clima/climacoupler-ci/builds/705#_). ClimaComms will be updated to 0.3.4 in #207 

Review checklist

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: experiments/AMIP/moist_mpi_earth/coupler_driver_modular.jl [hunk 1]
```diff
@@ -71,7 +72,7 @@ if isinteractive()
     parsed_args["vert_diff"] = true #hide
     parsed_args["rad"] = "gray" #hide
     parsed_args["microphy"] = "0M" #hide
-    parsed_args["energy_check"] = true
+    parsed_args["energy_check"] = false
```
Comment (juliasloan25) [inline experiments/AMIP/moist_mpi_earth/coupler_driver_modular.jl:75]: Do we need `#hide` here?

File: src/TimeManager.jl [hunk 2]
```diff
@@ -48,4 +49,18 @@ datetime_to_strdate(datetime::DateTime) =
...
+
+"""
+    trigger_callback(cs, ::Monthly)
+    
+Returns `true` if the current date is equal to or exceeds the saved first of the month at time of 00:00:00. 
+
+# Arguments
+- `cs`: [CoupledSimulation] containing info about the simulation
+"""
+trigger_callback(cs::CoupledSimulation, ::Monthly) = cs.dates.date[1] >= cs.dates.date1[1] ? true : false
```
Comment (juliasloan25) [inline src/TimeManager.jl:64]: why does this need the `::Monthly` param?

File: experiments/AMIP/moist_mpi_earth/user_diagnostics.jl [hunk 3]
```diff
@@ -0,0 +1,86 @@
+#import Diagnostics: get_var
+
+# Atmos diagnostics
+
+"""
+    get_var(cs, ::Val{:T})
```
Comment (juliasloan25) [inline experiments/AMIP/moist_mpi_earth/user_diagnostics.jl:6]: add type to function definition in comment (for all `get_var` functions)

File: src/Diagnostics.jl [hunk 4]
```diff
@@ -0,0 +1,215 @@
+#=
```
Comment (juliasloan25) [inline src/Diagnostics.jl:1]: change `#=`, `=#` to `"""` per Valeria's recent docstring updates

File: src/Diagnostics.jl [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    date_m1 = date - Dates.Day(1) # obtain previous month
+    datetime = Dates.DateTime(Dates.yearmonth(date_m1)[1], Dates.yearmonth(date_m1)[2])
+    Dates.datetime2unix(datetime)
+end
+
+save_time_format(date::Dates.DateTime, ::EveryTimestep) = Dates.datetime2unix(date)
+
+
+"""
+    pre_save(::TimeMean, cs::CoupledSimulation, dg::DiagnosticsGroup
```
Comment (juliasloan25) [inline src/Diagnostics.jl:187]: missing closing `)` (and in line 196)

File: test/diagnostics_tests.jl [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    accumulate_diagnostics!,
+    save_diagnostics,
+    TimeMean,
+    DiagnosticsGroup,
+    pre_save,
+    post_save,
+    save_time_format
+
+
+FT = Float64
```
Comment (juliasloan25) [inline test/diagnostics_tests.jl:23]: could add loop `for FT in (Float32, Float64)` over all testsets to test both FTs

**Review by juliasloan25 — APPROVED** [review]
looks good! just minor comments :)

File: src/TimeManager.jl (hunk 2 above)
Comment (LenkaNovak) [reply src/TimeManager.jl:64]: Right now it's more like a label for clarity. But in the future, we may want to have callbacks for `Daily` or `EveryTimestep`. This callback interface will probably change though once we start thinking about more complex timestepping.

File: test/diagnostics_tests.jl (hunk 6 above)
Comment (LenkaNovak) [reply test/diagnostics_tests.jl:23]: Yeah, I thought about it too, but then saw that it's not really testing any fundamental dependencies on the FT, so I figured Float64 suffices.  Trying to cut down on unnecessary CI ;)

**Pull request merged**